Raft longevities jobs

aleksbykov opened the pull request:
Trello: https://trello.com/c/JGUC6Y7Q

Adding tests and jenkins jobs for initial testing of new feature - raft 

## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [x] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [x] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [x] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

enaydanov commented:
@aleksbykov, please, review precommit and unittests failures.

aleksbykov commented:
Precommit failed with not related issue:
```
23:01:04  ************* Module sdcm.nemesis
23:01:04  sdcm/nemesis.py:1:0: R0801: Similar lines in 2 files
23:01:04  ==sdcm.cluster_aws:[555:573]
23:01:04  ==sdcm.cluster_gce:[89:105]
```
I think you already fixed it .

The unit test failed to not related issue:
```
TestBaseVersion.test_enterpris

23:08:02  >       self.assertEqual(version_list, ['2021.1'])
23:08:02  E       AssertionError: Lists differ: ['2022.1'] != ['2021.1']
```
It could be related to new version.

fgelcer commented:
> Precommit failed with not related issue:
> 
> ```
> 23:01:04  ************* Module sdcm.nemesis
> 23:01:04  sdcm/nemesis.py:1:0: R0801: Similar lines in 2 files
> 23:01:04  ==sdcm.cluster_aws:[555:573]
> 23:01:04  ==sdcm.cluster_gce:[89:105]
> ```
> 
> I think you already fixed it .
> 
> The unit test failed to not related issue:
> 
> ```
> TestBaseVersion.test_enterpris
> 
> 23:08:02  >       self.assertEqual(version_list, ['2021.1'])
> 23:08:02  E       AssertionError: Lists differ: ['2022.1'] != ['2021.1']
> ```
> 
> It could be related to new version.

i have fixed this issue, if you rebase on top of latest master it will be gone

fruch commented on internal_test_data/minimal_test_case.yaml:
```diff
@@ -16,3 +16,5 @@ stress_read_cmd: ["cassandra-stress user profile=/tmp/c-s_profile_4mv_5queries.y
 instance_type_db: 'i3.large'
 ami_id_db_scylla: 'ami-06f919eb'
+
+nemesis_include_filter: [['topology_changes'],['schema_changes']]
```
can we rename this configuration to `nemesis_selector` or something similar, include and filter are kind of the opposites, and it's a confusing name.

roydahan replied:
We can, I didn't have any other good name.
But this PR isn't the place.

fruch replied:
I've seen this PR adding this to `docs/configuration_options.md` so I assumed it's a new configuration option this PR introduced. (didn't see it being used till now)

Now I see it's already in place.

fruch commented on sdcm/stress_thread.py:
```diff
@@ -119,10 +119,10 @@ def create_stress_cmd(self, node, loader_idx, keyspace_idx):
+            nodes = [n for n in self.node_list if n.dc_idx == loader_idx %
+                     3]  # make sure each loader is targeting on datacenter/region
+            nodes = nodes or self.node_list
+            stress_cmd += ' -node {}'.format(','.join([node.ip_address for node in nodes]))
```
this change we should do for all of the stress threads, i.e. we should always select all known nodes

fruch reviewed (approved):
LGTM

fruch commented:
@aleksbykov is this PR still needed ? Or those were merged in a different PR?

roydahan commented:
https://github.com/scylladb/scylla-cluster-tests/pull/4948

aleksbykov commented:
@roydahan @fruch @fgelcer @KnifeyMoloko @vponomaryov @soyacz could you review?

roydahan commented on test-cases/raft/raft-multidc-parallel-topology-schema-changes-12h.yaml:
```diff
@@ -0,0 +1,30 @@
+test_duration: 800
```
@aleksbykov I don't think these test cases should be under raft folder.
They are general longevities.
When we add the configuration to enable raft, they test it with raft.

roydahan commented on test-cases/raft/raft-parallel-schema-changes-12h.yaml:
```diff
@@ -0,0 +1,26 @@
+test_duration: 800
```
same as above.

roydahan commented on test-cases/raft/raft-schema-changes-3h.yaml:
```diff
@@ -0,0 +1,26 @@
+test_duration: 260
```
same as above.

roydahan commented on test-cases/raft/raft-topology-changes-3h.yaml:
```diff
@@ -0,0 +1,26 @@
+test_duration: 260
```
same

roydahan commented on test-cases/raft/raft-topology-changes-double-cluster-size-12h.yaml:
```diff
@@ -0,0 +1,26 @@
+test_duration: 800
```
same + the name of this test isn't clear.
What does it mean changes-double-cluster-size?
Pick a better name.

roydahan commented on jenkins-pipelines/feature-raft-double-cluster-size-12h.jenkinsfile:
```diff
@@ -0,0 +1,13 @@
+    backend: 'aws',
+    region: 'eu-west-1',
+    test_name: 'longevity_test.LongevityTest.test_custom_time',
+    test_config: '''["test-cases/raft/raft-topology-changes-double-cluster-size-12h.yaml", "configurations/raft/enable_raft_experimental.yaml"]''',
```
Change all locations of test config not being under "raft" folder.

roydahan reviewed (changes requested):
First review, please change locations of files not being under "Raft" folder.

fruch commented on sdcm/nemesis.py:
```diff
@@ -1687,15 +1696,24 @@ def build_list_of_disruptions_to_execute(self, nemesis_selector=None, nemesis_mu
-        self.log.debug("This is the list of callable disruptions {}".format(self.disruptions_list))
         return self.disruptions_list
 
     @property
     def nemesis_selector_list(self) -> list:
+        if self._nemesis_selector_list:
+            return self._nemesis_selector_list
```
cause it's using a `@nemesis_selector_list.setter`, i'm not sure it could work together as expected

fruch commented on unit_tests/test_nemesis.py:
```diff
@@ -223,3 +226,214 @@ def test_use_disabled_monkey():
     assert 'disrupt_toggle_table_gc_mode' in collected_disrupt_methods_names
+
+
+class TestSisyphusMonkeyNemesisFilter:
```
I agree with @soyacz we should keep local list like that 
we have tests that tests the filters, we don't need to test each group we have is exactly something we expect.

we might want a tool/test that would help us get the list base on some specific filter (for planning and such)
but not like this.

aleksbykov replied:
I removed the most largest categories. i would like to left check of other categories, to be sure that even after new nemesis added to such category the test catch such changes

fruch commented on sdcm/sct_config.py:
```diff
@@ -1361,7 +1361,7 @@ class SCTConfiguration(dict):
              help="Flag for running db node benchmarks before the tests"),
         dict(name="nemesis_selector", env="SCT_NEMESIS_SELECTOR",
-             type=str_or_list, k8s_multitenancy_supported=True,
+             type=str_or_list_or_eval, k8s_multitenancy_supported=True,
```
why you need it eval ? in which use case ?

aleksbykov replied:
In case when parameter will taken from env variable SCT_NEMESIS_SELECTOR

fruch commented on sdcm/cluster.py:
```diff
@@ -4189,8 +4189,12 @@ def wait_total_space_used_per_node(self, size=None, keyspace='keyspace1'):
                 self.nemesis.append(nem['nemesis'](tester_obj=tester_obj,
-                                                   termination_event=self.nemesis_termination_event))
+                                                   termination_event=self.nemesis_termination_event),
+                                    nemesis_selector=nem['nemesis_selector'])
```
That's true, but for one `CategoricalMonkey` it's not

soyacz replied:
you fixed CategoricalMonkey but all other nemesis that don't override `__init__` will also fail. I think we need to add also **kwargs to base `Nemesis` class `__init__`

aleksbykov commented:
> First review, please change locations of files not being under "Raft" folder.

@roydahan , i moved the config files and renamed them. Does they look clearly now?

roydahan commented on test-cases/longevity/longevity-change-cluster-size-by-2-times.yaml:
```diff
@@ -0,0 +1,26 @@
+
+seeds_num: 3
+
+user_prefix: 'raft-double-cluster-size-with-schema-changes-12h'
```
Why is it called "-with-schema-changes"?
Where are the schema changes?

aleksbykov replied:
fixed with nemesis filter

roydahan commented on test-cases/longevity/longevity-multidc-parallel-topology-schema-changes-12h.yaml:
```diff
@@ -0,0 +1,30 @@
+test_duration: 800
+
+prepare_write_cmd:  "cassandra-stress write cl=QUORUM n=20971520 -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=80 -pop seq=1..20971520 -col 'n=FIXED(10) size=FIXED(512)' -log interval=5"
```
Don't you need to set here something like this:

'replication(strategy=NetworkTopologyStrategy,us-east1scylla_node_east=3,us-west1scylla_node_west=3)'

aleksbykov replied:
fixed

roydahan commented on jenkins-pipelines/feature-raft-double-cluster-size-12h.jenkinsfile:
```diff
@@ -0,0 +1,13 @@
+#!groovy
```
Can you please put this pipeline file under "raft" folder?

roydahan replied:
Also, I don't think it should be called "feature-raft".
The scenario is valid also without raft.

And we should have a pipeline also without the experimental raft for the regular longevities.

aleksbykov replied:
renamed

roydahan commented on jenkins-pipelines/feature-raft-multidc-schema-topology-changes-12h.jenkinsfile:
```diff
@@ -0,0 +1,13 @@
+#!groovy
```
same

aleksbykov replied:
renamed

roydahan commented on jenkins-pipelines/feature-raft-schema-changes-3h.jenkinsfile:
```diff
@@ -0,0 +1,13 @@
+#!groovy
```
same as above.

aleksbykov replied:
renamed

roydahan commented on jenkins-pipelines/feature-raft-schema-topology-changes-12h.jenkinsfile:
```diff
@@ -0,0 +1,13 @@
+#!groovy
```
same as above.

aleksbykov replied:
renamed

roydahan commented on jenkins-pipelines/feature-raft-topology-changes-3h.jenkinsfile:
```diff
@@ -0,0 +1,13 @@
+#!groovy
```
same as above

aleksbykov replied:
renamed

roydahan reviewed (changes requested):
More comments (still didn't finish review)

aleksbykov commented:
@fruch @soyacz Could you review again?

roydahan commented:
@aleksbykov do you have these longevities on the test plan or a place with the reasons of why to run each of these longevities?
(e.g. why do we need to run the doucble-cluster-size?)

Also, I want to add more tests around multidc, NetworkMonkey and in parallel to schema_changes / topology_changes.

aleksbykov commented:
@roydahan , It was discussed for raft feature testing: https://docs.google.com/document/d/15J4Ha2OHyvXh6hvFXDMKJhyf3s4-LR-lxxG84Dv-u9Y/edit#heading=h.5kw1yomclkji.

fruch commented:
@aleksbykov 

unittest are failing 
```
16:35:23  FAILED unit_tests/test_utils__operator__multitenant_common.py::UtilsOperatorMultitenantCommonTests::test_multitenant_longevity_class_with_unique_options
16:35:23  FAILED unit_tests/test_utils__operator__multitenant_common.py::UtilsOperatorMultitenantCommonTests::test_multitenant_performance_class_with_unique_options
```

fruch commented:
also the test case linting is failing:
https://jenkins.scylladb.com/job/sct-github-PRs-scan/job/scylla-cluster-tests/job/PR-4453/33/execution/node/49/log/

soyacz commented on unit_tests/test_nemesis.py:
```diff
@@ -223,3 +226,124 @@ def test_use_disabled_monkey():
+            "disrupt_toggle_cdc_feature_properties_on_table",
+        ]
+
+    def test_list_topology_changes_monkey(self, expected_topology_changes_methods):
```
test with such name exists (line 175) - and looks similar.

aleksbykov replied:
Yes, it was duplicated. Removed one

roydahan commented on test-cases/longevity/longevity-multidc-parallel-topology-schema-changes-12h.yaml:
```diff
@@ -0,0 +1,30 @@
+server_encrypt: true
+internode_encryption: 'dc'
+
+ip_ssh_connections: 'public'
```
Now we don't need public, we have vpc_peering.

roydahan commented on test-cases/longevity/longevity-multidc-parallel-topology-schema-changes-12h.yaml:
```diff
@@ -0,0 +1,30 @@
+             ]
+
+n_db_nodes: '4 4'
+n_loaders: '2 2'
```
I think you should change to '1 1'.
Also add round_robin.

roydahan commented on test-cases/longevity/longevity-multidc-parallel-topology-schema-changes-12h.yaml:
```diff
@@ -0,0 +1,30 @@
+
+prepare_write_cmd:  "cassandra-stress write cl=QUORUM n=20971520 -schema 'replication(strategy=NetworkTopologyStrategy,eu-west1scylla_node_west1=3,eu-west2scylla_node_west2=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=80 -pop seq=1..20971520 -col 'n=FIXED(10) size=FIXED(512)' -log interval=5"
+
+stress_cmd: ["cassandra-stress write cl=QUORUM duration=720m -schema 'replication(strategy=NetworkTopologyStrategy,eu-west1scylla_node_west=3,eu-west2scylla_node_west=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=40 -pop 'dist=uniform(1..20971520)' -col 'n=FIXED(10) size=FIXED(512)' -log interval=5",
```
Add a c-s user profile with MV (simple one that builds a view on the main table keyspace1.standard1".

roydahan commented on test-cases/longevity/longevity-multidc-parallel-topology-schema-changes-12h.yaml:
```diff
@@ -0,0 +1,30 @@
+region_aware_loader: true
+n_monitor_nodes: 1
+
+instance_type_db: 'i3.4xlarge'
```
let's use i3en.xlarge or even i4i.xlarge (if it's available in these regions).

roydahan commented on test-cases/longevity/longevity-multidc-parallel-topology-schema-changes-12h.yaml:
```diff
@@ -0,0 +1,30 @@
+ip_ssh_connections: 'public'
+intra_node_comm_public: true
+
+user_prefix: 'longevity-multidc-12h'
```
Change this to be something more descriptive.

roydahan commented on test-cases/longevity/longevity-change-cluster-size-by-2-times.yaml:
```diff
@@ -0,0 +1,27 @@
+test_duration: 800
+
+prepare_write_cmd:  "cassandra-stress write cl=QUORUM n=20971520 -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=80 -pop seq=1..20971520 -col 'n=FIXED(10) size=FIXED(512)' -log interval=5"
+
+stress_cmd: ["cassandra-stress write cl=QUORUM duration=720m -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=40 -pop 'dist=uniform(1..20971520)' -col 'n=FIXED(10) size=FIXED(512)' -log interval=5",
+             "cassandra-stress read  cl=QUORUM duration=720m -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=40 -pop 'dist=uniform(1..20971520)' -col 'n=FIXED(10) size=FIXED(512)' -log interval=5"
+             ]
+
+n_db_nodes: 4
+n_loaders: 2
+n_monitor_nodes: 1
+
+instance_type_db: 'i3.4xlarge'
+
+nemesis_class_name: 'GrowShrinkClusterNemesis:1 SisyphusMonkey:1'
+nemesis_selector: [[],["schema_changes"]]
+nemesis_interval: 5
+nemesis_filter_seeds: false
+nemesis_add_node_cnt: 4
+
+seeds_num: 3
+
+user_prefix: 'raft-double-cluster-size-with-schema-changes-12h'
```
Same comments as other tests below.

roydahan commented on test-cases/longevity/longevity-schema-changes-3h.yaml:
```diff
@@ -0,0 +1,26 @@
+test_duration: 260
+
+prepare_write_cmd:  "cassandra-stress write cl=QUORUM n=20971520 -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=80 -pop seq=1..20971520 -col 'n=FIXED(10) size=FIXED(512)' -log interval=5"
+
+stress_cmd: ["cassandra-stress write cl=QUORUM duration=180m -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=40 -pop 'dist=uniform(1..20971520)' -col 'n=FIXED(10) size=FIXED(512)' -log interval=5",
+             "cassandra-stress read  cl=QUORUM duration=180m -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=40 -pop 'dist=uniform(1..20971520)' -col 'n=FIXED(10) size=FIXED(512)' -log interval=5"
+             ]
```
In this longevity it's fine to keep it like this.

roydahan commented on test-cases/longevity/longevity-schema-changes-3h.yaml:
```diff
@@ -0,0 +1,26 @@
+n_loaders: 2
+n_monitor_nodes: 1
+
+instance_type_db: 'i3.4xlarge'
```
Change to i3en.large.

roydahan commented on test-cases/longevity/longevity-schema-changes-3h.yaml:
```diff
@@ -0,0 +1,26 @@
+space_node_threshold: 64424
+
+pre_create_schema: True
+sstable_size: 100
```
What's that?

roydahan commented on test-cases/longevity/longevity-topology-changes-3h.yaml:
```diff
@@ -0,0 +1,26 @@
+test_duration: 260
+
+prepare_write_cmd:  "cassandra-stress write cl=QUORUM n=20971520 -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=80 -pop seq=1..20971520 -col 'n=FIXED(10) size=FIXED(512)' -log interval=5"
+
+stress_cmd: ["cassandra-stress write cl=QUORUM duration=180m -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=40 -pop 'dist=uniform(1..20971520)' -col 'n=FIXED(10) size=FIXED(512)' -log interval=5",
+             "cassandra-stress read  cl=QUORUM duration=180m -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=40 -pop 'dist=uniform(1..20971520)' -col 'n=FIXED(10) size=FIXED(512)' -log interval=5"
+             ]
```
Use s-b large partitions but simple one.

roydahan commented on test-cases/longevity/longevity-topology-changes-3h.yaml:
```diff
@@ -0,0 +1,26 @@
+n_loaders: 2
+n_monitor_nodes: 1
+
+instance_type_db: 'i3.4xlarge'
```
Let's use asymmetric cluster.
Change to i3en.2xlarge.

roydahan commented on test-cases/longevity/longevity-topology-changes-3h.yaml:
```diff
@@ -0,0 +1,26 @@
+nemesis_interval: 5
+nemesis_filter_seeds: false
+
+seeds_num: 2
```
Let's use 4.

roydahan commented on test-cases/longevity/longevity-topology-changes-3h.yaml:
```diff
@@ -0,0 +1,26 @@
+space_node_threshold: 64424
+
+pre_create_schema: True
+sstable_size: 100
```
?

aleksbykov commented:
@roydahan could you review again.
@vponomaryov , Could you check multitenant unit tests? I fixed them to pass, but it will good to know that the logic is not broken

vponomaryov commented on unit_tests/test_utils__operator__multitenant_common.py:
```diff
@@ -177,12 +177,12 @@ def _multitenant_class_with_unique_options(self, klass):
         for cmd_name in ("stress_cmd_r", "stress_cmd_m"):
             self.assertEqual(
-                tenants[0].params.get(cmd_name),
+                tenants[0].params.get(cmd_name)[0],
```
The fact that you need to change this file means that the K8S is broken for sure.

vponomaryov reviewed (changes requested):
@aleksbykov 
Please see the inline comments. Also, general comment:

Why do you need all these config changes knowing that the same options are already able to serve the multiple usages like it is done in the `K8S multitenant` setups.

Here is how we covered the same use case in K8S:
- https://github.com/scylladb/scylla-cluster-tests/blob/master/longevity_operator_multi_tenant_test.py
- https://github.com/scylladb/scylla-cluster-tests/blob/master/sdcm/utils/operator/multitenant_common.py

Above is pretty small changes.

What blocks you from reusing similar approach?
Maybe it is easier to create LongevityV2?

aleksbykov commented on sdcm/cluster.py:
```diff
@@ -4217,8 +4217,12 @@ def wait_total_space_used_per_node(self, size=None, keyspace='keyspace1'):
     def add_nemesis(self, nemesis, tester_obj):
         for nem in nemesis:
             for _ in range(nem['num_threads']):
+                nemesis_obj = nem['nemesis'](tester_obj=tester_obj,
+                                             termination_event=self.nemesis_termination_event)
+                nemesis_obj.nemesis_selector_list = nem['nemesis_selector']
```
fixed

aleksbykov commented on sdcm/sct_config.py:
```diff
@@ -70,7 +70,7 @@ def str_or_list(value: Union[str, List[str], List[List[str]]]) -> List[str]:
     if isinstance(value, list):
         for element in value:
-            if isinstance(element, str):
+            if isinstance(element, (str, list)):
```
removed not needed changes

aleksbykov commented:
@vponomaryov thank you very much for your comments. I fixed inline comments.
The general idea was to allow run several SisyphusMonkeys in parallel in same job with different nemesis_selector.

aleksbykov commented:
> @aleksbykov Please see the inline comments. Also, general comment:
> 
> Why do you need all these config changes knowing that the same options are already able to serve the multiple usages like it is done in the `K8S multitenant` setups.

As i wrote above it is mostly needs to run in parallel several nemssises with different selectors. I was not familiar with multitenant  code, when prepare the changes. Now i checked, and:
1. multitenant doesn't support parallel nemesis run per tenant?

> 
> Here is how we covered the same use case in K8S:
> 
> * https://github.com/scylladb/scylla-cluster-tests/blob/master/longevity_operator_multi_tenant_test.py
> * https://github.com/scylladb/scylla-cluster-tests/blob/master/sdcm/utils/operator/multitenant_common.py
> 
> Above is pretty small changes.
> 

> What blocks you from reusing similar approach? Maybe it is easier to create LongevityV2?

I don't want to create more test instances, while needs only correctly parse the config parameter

Also i have a question, why could unit test failed for stress command, if i didn't touch the code related to stress

vponomaryov commented on sdcm/cluster.py:
```diff
@@ -4217,8 +4217,10 @@ def wait_total_space_used_per_node(self, size=None, keyspace='keyspace1'):
-                                                   termination_event=self.nemesis_termination_event))
+                nemesis_obj = nem['nemesis'](tester_obj=tester_obj,
+                                             termination_event=self.nemesis_termination_event)
+                nemesis_obj.nemesis_selector_list = nem['nemesis_selector']
```
Why do we set it separately after the init completed and not as part of the `init`? Any problem with it?
I would expect addition of one more parameter call here like following:

```
self.nemesis.append(nem['nemesis'](
    tester_obj=tester_obj,
    termination_event=self.nemesis_termination_event,
    nemesis_selector_list=nem['nemesis_selector'],
)
```

aleksbykov replied:
Because we didn't call it in Nemesis constructor, and run it only in specific nemesis Sisyphus. But i think __init__ could be changes a  bit to support that

vponomaryov replied:
So, do you plan to do it in scope of this PR?

vponomaryov replied:
This is still actual.

aleksbykov replied:
Fixed

vponomaryov commented on unit_tests/test_utils__operator__multitenant_common.py:
```diff
@@ -197,7 +197,7 @@ def _multitenant_class_with_unique_options(self, klass):
         # Process tenant-2 specific options
         self.assertEqual(
-            tenants[1].params.get("stress_cmd_r"), "fake__stress_cmd_r__tenant2__single_str_in_list")
+            tenants[1].params.get("stress_cmd_r"), ["fake__stress_cmd_r__tenant2__single_str_in_list"])
```
It must not be changed, why the `stress_cmd_r` option is expected to behave differently here?

aleksbykov replied:
yes, if fixed it already

aleksbykov commented:
@vponomaryov , i finally fixed, so now multitenant and regular parallel nemesis tests are passed. Could you please review one  more time?

vponomaryov commented on sdcm/sct_config.py:
```diff
@@ -1845,7 +1845,9 @@ def _validate_value(self, opt):
                     and opt.get("k8s_multitenancy_supported")
                     and isinstance(self.get(opt['name']), list)):
-                raise ValueError("failed to validate {}".format(opt['name'])) from ex
+                if not (isinstance(self.get('nemesis_class_name'), str)
+                        and len(self.get("nemesis_class_name").split(" ")) > 1):
```
Actually, you have bug here.
If `nemesis_class_name` is set to be string and has multiple words separated with space then any other broken option won't be caught.

It must look like something following:
```
if not (
    opt['name'] == 'nemesis_class_name' 
    and isinstance(self.get(opt['name']), str) 
    and len(self.get("nemesis_class_name").split(" ")) > 1):
```

aleksbykov replied:
fixed

vponomaryov commented on unit_tests/test_nemesis.py:
```diff
@@ -12,9 +13,11 @@
 from sdcm.cluster_gce import ScyllaGCECluster
 from sdcm.cluster_aws import ScyllaAWSCluster
 from sdcm.cluster_docker import ScyllaDockerCluster
+from .test_tester import ClusterTesterForTests
```
Do not use relative imports. It may easily become a problematic point.

aleksbykov replied:
fixed

aleksbykov commented:
@vponomaryov could you look one more time?

vponomaryov commented:
> @vponomaryov could you look one more time?

Now it looks ok to me. Need to run some tests with the latest change.
But before it, need to solve comments made by others on the config files.

roydahan commented on test-cases/longevity/longevity-topology-changes-3h.yaml:
```diff
@@ -0,0 +1,30 @@
+seeds_num: 2
+
+
+user_prefix: 'longevity-topologychanges-async-3h'
```
What does "async" mean here?

aleksbykov replied:
It means asymetric shards. Lost flag. Renamed and add config parameter

roydahan commented on jenkins-pipelines/longevity-double-cluster-with-schema-changes-with-raft.jenkinsfile:
```diff
@@ -0,0 +1,13 @@
+    backend: 'aws',
+    region: 'eu-west-1',
+    test_name: 'longevity_test.LongevityTest.test_custom_time',
+    test_config: '''["test-cases/longevity/longevity-change-cluster-size-by-2-times.yaml", "configurations/raft/enable_raft_experimental.yaml"]''',
```
Don't we need also to change now the "raft" to official flag or in addition to it?

aleksbykov replied:
The new flag `consistent_cluster_management: true` will enable on 5.2 the functionality which was enabled by `experimenta_feature: raft`. The new flag will be enabled by default for new clusters according to https://github.com/scylladb/scylladb/issues/12470. 
the experimental flag `experimenta_feature: raft` will be used for next versions (master, 5.3) for other raft features

New flag consistent_cluster_management is already in scylla_yaml.py. I can switch default value for this parameter to True in new commit in this PR

roydahan replied:
ok, so I'm not sure if for these longevities we need the raft experimental flag or not.

aleksbykov replied:
For 5.2 it may be not need, but for master and for next version it could be used.

roydahan commented on test-cases/longevity/longevity-topology-changes-3h.yaml:
```diff
@@ -0,0 +1,32 @@
+n_loaders: 2
+n_monitor_nodes: 1
+
+instance_type_db: 'i3.2xlarge'
```
In this case, that we use random number of shards, I think it worth changing the instance type to have more shards.
I think even i4i.8xlarge.

aleksbykov replied:
changed

roydahan commented on jenkins-pipelines/longevity-double-cluster-with-schema-changes-with-raft.jenkinsfile:
```diff
@@ -0,0 +1,13 @@
+#!groovy
```
Let's move these under raft folder as well.

aleksbykov replied:
done

roydahan commented:
Will rebase 5.2 so it gets in.

roydahan commented:
Please run the tool that creates the jobs for master, 5.2 and 2023.1

roydahan commented:
I ran the job to create all the new jobs in 5.2 & 2023.1
